src/framework/components/rigid-body/system.js

```diff
@@ -552,6 +560,242 @@ class RigidBodyComponentSystem extends ComponentSystem {
+        Debug.assert(Ammo.ConcreteContactResultCallback, 'pc.RigidBodyComponentSystem#shapeCast: Your version of ammo.js does not expose Ammo.ConcreteContactResultCallback. Update it to latest.');
+
+        if (shape.type === 'capsule') {
+            return this.capsuleCast(shape.radius, shape.height, shape.axis, position, rotation);
+        } else if (shape.type === 'cone') {
+            return this.coneCast(shape.radius, shape.height, shape.axis, position, rotation);
+        } else if (shape.type === 'cylinder') {
+            return this.cylinderCast(shape.radius, shape.height, shape.axis, position, rotation);
+        } else if (shape.type === 'sphere') {
+            return this.sphereCast(shape.radius, position, rotation);
+        }
+
+        return this.boxCast(shape.halfExtents, position, rotation);
+    }
+
+    /**
+     * Perform a collision check on the world and return all entities the box hits.
+     * It returns an array of {@link RaycastResult}, one for each hit. If no hits are
+     * detected, the returned array will be of length 0.
+     *
+     * @param {Vec3} halfExtents - The half-extents of the box in the x, y and z axes.
+     * @param {Vec3} position - The world space position for the box to be.
+     * @param {Vec3} rotation - The world space rotation for the box to have.
+     *
+     * @returns {RaycastResult[]} An array of boxcast hit results (0 length if there were no hits).
+     */
+    boxCast(halfExtents, position, rotation) {
+        Debug.assert(Ammo.ConcreteContactResultCallback, 'pc.RigidBodyComponentSystem#boxCast: Your version of ammo.js does not expose Ammo.ConcreteContactResultCallback. Update it to latest.');
+
+        ammoVec3.setValue(halfExtents.x, halfExtents.y, halfExtents.z);
+        return this._shapecast(new Ammo.btSphereShape(ammoVec3), position, rotation);
+    }
+
+    /**
+     * Perform a collision check on the world and return all entities the capsule hits.
+     * It returns an array of {@link RaycastResult}, one for each hit. If no hits are
+     * detected, the returned array will be of length 0.
+     *
+     * @param {number} radius - The radius of the capsule.
+     * @param {number} height - The total height of the capsule from tip to tip.
+     * @param {number} axis - The local space axis with which the capsule's length is aligned. 0 for X, 1 for Y and 2 for Z. Defaults to 1 (Y-axis).
+     * @param {Vec3} position - The world space position for the capsule to be.
+     * @param {Vec3} rotation - The world space rotation for the capsule to have.
+     *
+     * @returns {RaycastResult[]} An array of capsulecast hit results (0 length if there were no hits).
+     */
+    capsuleCast(radius, height, axis, position, rotation) {
+        Debug.assert(Ammo.ConcreteContactResultCallback, 'pc.RigidBodyComponentSystem#capsuleCast: Your version of ammo.js does not expose Ammo.ConcreteContactResultCallback. Update it to latest.');
+
+        if (axis === 0) {
```

Comment:
I would try to keep variables of constant type .. you pass in a number, and here you assign string to it. I'd prefer you use new variable for string instead. I would probably just have an array with these 3 strings as a global array variable, and simply index into it here. And assert if the lookup is undefined.

Reply:
Ah, yeah. I'd agree with that. Best to keep the type of a variable the same as it flows through the code.

Reply:
Thanks for the feedback, it's fixed with last commit!